### ISSUE TITLE
[Suggestion] dataFrom for a directory

### ISSUE THREAD

**Issue by Aransh** [external-secrets/external-secrets#2987]
**Is your feature request related to a problem? Please describe.**
I manage the secrets in my Hashicorp Vault cluster using Terraform (for example creating AWS access keys, and then writing them to a vault secret).
My logical Vault layout is:
<cluster_name>/<application_name>/<secret_type>

Now, since I have multiple secret types, I must manage them separately (as I can't push them all in a single secret since Terraform will overwrite the data)
And since I want to inject all secret types of an application into it, I must manage an "externalSecret" resource per secret type (each with reference to the specific secret type name), and manually enable/disable secret types per application (as most applications don't require all secret types, and creating an ExternalSecret pointing to a non-existing secret will make it appear unhealthy).

TLDR - I have multiple secrets in the same directory, I want them all to dynamically get the same treatment (be injected into the same k8s secret), but have to manually specify each secret in my implementation.

**Describe the solution you'd like**
Best approach in my opinion, would be for "dataFrom" to support directories (instead of only final keys).
If provided path is a directory and not a key, get data from all secrets in this directory (maybe also add a "recursive" option?)

Example:
```
# Will pull data from all secrets under "test" directory
spec:
  dataFrom:
    - extract:
        key: test/

# Will pull data from all secrets under test/object-storage secret (same as current behavior)
spec:
  dataFrom:
    - extract:
        key: test/object-storage
```
**Describe alternatives you've considered**
Possibly regex would be a solution - https://external-secrets.io/latest/guides/getallsecrets/

**Comment by Aransh** [comment]
Update, tried the regex solution, and while it's a step forward, it doesn't solve my problem, as it just adds any matching secret as a single key (while I need all keys from all matching secrets to be merged)

**Comment by Skarlso** [comment]
Hello.

Maybe I'm too new here. :D But could you please define what you mean by directory / folder? I assume you don't mean a physical folder, right? But some kind of namespace derived location?

**Comment by Aransh** [comment]
@Skarlso Correct, not really a namespace, this is comparable to a directory in a filesystem.
Don't know how this works in other secret managers external-secrets supports, but in Hashicorp Vault the data is structured in directories.
So for example I can have a secret called "object-storage", it's content will be key-value, as followed:
```
ACCESS_KEY=xxxx
SECRET_KEY=yyy
```
This secret is related to a specific application, let's say "app1", which is deployed under my cluster called "cluster1".
So I'll place this secret under directory path "cluster1/app1/"

Finally, in external secrets, I will point to a secret stored under "cluster1/app1/object-storage", and my secret will contain the actual key-values inside the secret itself.

What I'm asking for is the ability to just set the directory, e.g, "cluster1/app1/" instead of the key, which will pull the key-values from all secrets inside this "directory" (path), including "object-storage", and any other secrets that may be present there, without me having to hardcode the name of each secret I create in this directory.

Note that pulling daya from several secrets is already supported, it's just not dynamic, since I need to specify all secret names.

**Comment by Skarlso** [comment]
Interesting. And you're saying that with regex you could specify a glob or something but that didn't work because it will create a single secret containing all of those items.

**Comment by Skarlso** [comment]
This sounds almost like this https://github.com/external-secrets/external-secrets/issues/2670 but this talks about multiple external secrets.

However the post from Moritz actually explains why a single ES is doing what it does.

Basically, TL;DR an ExternalSecrets manages a single Secret because it merges keys and does reconciliation when things are changing. If it would track multiple secrets there would be no nice way to design a diff, or tracking what status changes, where and why.

https://github.com/external-secrets/external-secrets/issues/2670#issuecomment-1701740231

**Comment by Aransh** [comment]
You are correct, this is exactly what I'm requesting, couldn't find it myself.
I'll close this one then and add my 2 cents there, thanks

**Comment by Aransh** [comment]
@Skarlso  just took a second look, and I seem to have misunderstood the linked issue.
They are requesting creating multiple secrets from all secrets in a directory
I am requesting creating a single, merged secret from all secrets in a directory (which seems to me aligns more with the existing external-secrets feature set).

Re-opening this

**Comment by Skarlso** [comment]
Oh wait, so why is the regex thing not working then? You should be able to create a single secret with all the data.

>Update, tried the regex solution, and while it's a step forward, it doesn't solve my problem, as it just adds any matching secret as a single key (while I need all keys from all matching secrets to be merged)

You should be able to create those. Try using Templates with rewrite rules?

**Comment by Skarlso** [comment]
can you take a look at https://external-secrets.io/latest/guides/datafrom-rewrite/?

**Comment by Aransh** [comment]
@Skarlso well, unless I'm missing something here, what trying this gave me is not combining the keys/values from each secret, but just using the actual secret name as the key, and the key/values as the values.

e.g. I have 2 secrets, one is called "object-storage" and has this content:
`{
"ACCESS_KEY": "XXXX"
"SECRET_KEY": "YYYY"
}`

The other for called "mongo-credentials" with this content:

`{
"USERNAME": "XXXX"
"PASSWORD": "YYYY"
}`

I want to provide all these secrets to my app (as they are stored on a specific path dedicated to it), so I'd like regexp to combine them into a k8s secret with this structure:

`{
"ACCESS_KEY": "XXXX"
"SECRET_KEY": "YYYY"
"USERNAME": "XXXX"
"PASSWORD": "YYYY"
}`

But instead, using regexp with rewrite (to remove the path from the name), I get this final secret:

`{
"object-storage": {
  "ACCESS_KEY": "XXXX"
  "SECRET_KEY": "YYYY"
}
"mongo-credentials": {
  "USERNAME": "XXXX"
  "PASSWORD": "YYYY"
}`

Which is not as usable, as it's not what my application expects...

So instead I'm forced to create 2 separate externalSecrets with "extract", one for each secret (and one more/less for any future secret I might add/remove from this path)

**Comment by Skarlso** [comment]
Huh. Interesting. I honestly thought the former would happen. Like it should just merge all keys without adding stuff to it. 

I'll check this out when I'm back home today.

**Comment by Skarlso** [comment]
Yah, the code pretty much puts all of the found secrets into their respective name based location.

I didn't find a way to do what you want. :(

**Comment by Aransh** [comment]
@Skarlso I feel like the solution I'm looking for should basically be a combination of "extract" and "find"

**Comment by Skarlso** [comment]
Damn, this is what I was missing:

```yaml
    rewrite:
    - merge: {} # <- merge all secret contents into a "combined" key
```

I tried using Path. :D This feature is a bit obscure.

**Comment by moolen** [comment]
The `merge: {}` doesn't exist yet, it's just an idea :D

**Comment by Skarlso** [comment]
I  👏  completely  👏  misread 👏  everything. 👏

**Comment by Aransh** [comment]
Hey @moolen 😄
Thanks for the detailed response!

I did find the "path" option, but it doesn't do much I can't already do without it.

The "literal" option seems interesting, definitely wouldn't have found it myself, but also looks way too complicated (I wouldn't want to leave this code for anyone else to maintain 😅)

The additional rewrite options idea sounds exactly like what I was thinking.

I will add that while this feature will significantly simplify my infrastructure, development, and deployment processes, it is not a blocker.
Just thought it might be a great feature that could help other people as well, so I figured I'd raise it here 😁

So if you'd decide to implement it one day, I'd be the first to use it.

Also one last suggestion (again, not a must, but nice to have):
As I mentioned, my use case is several secrets created by different processes (which is why they are separate secrets to begin with), and I also always add one "overwrite" secret, where I give devs write permissions to be able to add any additional secrets they may want, and also be able to manually overwrite values from the other "generated" secrets which they can't edit.
For that use case, if a "merge" option were to be added, it would be great to be able to specify priority for specific secret(s), if they exist
I'm thinking maybe something like:


`apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: example
spec:
  # ...
  dataFrom:
  - find:
      path: cluster/application
      name: 
        regexp: ".*"
    rewrite:
    - merge: {}
      priority:
      - overwrite
      - overwrite_also`

In this case, the secret named "overwrite" will take precedence over all other secrets in case of key duplication (only if it exists), followed by "overwrite_also", and only then all the other found secrets (which will probably just prioritize alphabetically in case of key duplication)

Anyway, just my 2 cents, thanks again for the response 🙏🏼

**Comment by moolen** [comment]
> I will add that while this feature will significantly simplify my infrastructure, development, and deployment processes, it is not a blocker.
Just thought it might be a great feature that could help other people as well, so I figured I'd raise it here 😁

:heart: much appreciated! Thank you for providing feedback and sharing your perspective, it's hard to get feedback from users :D

I'm afraid we have to rely on contributions from the community to deliver this new feature, i've added `good first issue` which hopefully gives it a bit more visibility.


Let me try to sketch this out a bit further. I probably went a bit too far, but at it's core that's roughly how i think this feature could work.

**tl;dr**

with these secrets
```sh
cluster1/backend/mysql-creds # {db_user, db_pass}
cluster1/backend/aws-creds # {AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY}
cluster1/backend/custom-stuff-controlled-by-developers # {foo, bar}
```
 
We want a secret that looks like this:

```json
{
  "db_user": "****",
  "db_pass": "****",
  "AWS_ACCESS_KEY_ID": "****",
  "AWS_SECRET_ACCESS_KEY": "****",
  "foo": "****",
  "bar": "****"
}
```

I though through this an looked at the code, i think this structure should be implementable and should work for your use-case.

```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: example
spec:
  # ...
  dataFrom:
  - find:
      path: cluster1/backend
      name: 
        regexp: ".*"
    rewrite:
    - merge:
       # define the key under which we want to store the result,
       # this contains the merged value.
       # Must be omitted with mergeStrategy=Extract
        into: ""

        # We deal with map[string][]byte under the hood,
        # hence we need a strategy to merge the bytes. 
        #
        # JSON: keep the keys (secret names from the provider)
        # and only merge the secret values, treating it as JSON.
        # The resulting JSON object is stored in the `into` property.
        #
        # Template: make use of the template engine to merge
        # the values (TBD).
        #
        # Extract: parse the value as JSON and hoist
        # the key/value pairs up just like dataFrom.Extract it does.
        strategy: Extract

        # When there is a conflict when merging the []byte,
        # e.g. there is a key AWS_ACCESS_KEY that already exists
        # which would be overridden, we need a way to resolve
        # that conflict: either bail out or ignore.
        # Merging the secrets from the priority field below
        # is not considered a conflict and will never trigger a conflictPolicy error.
        conflictPolicy: Ignore # or ErrorOut

        # when we merge the structure map[string][]byte
        # we need to define a stable sort order in which we apply the merge.
        # This is only relevant for conflictPolicy=Ignore.
        sort: 
          strategy: Numeric
          order: Ascending

        # define a list of keys that take precedence over all keys,
        # no matter what sorting strategy/order is being used.
        # These priority-keys are merged into the result map
        # in the defined order.
        # A key conflict will not trigger an error (as per conflictPolicy), instead
        # the values is simply overridden.
        priority:
          - custom-stuff-controlled-by-developers
```

We need to take care of a couple of things to guarantee a stable sort order and provide extensibility for other merging strategies.

When we make the assumption that `conflictPolicy=ErrorOut` (which should be the default) then we don't need to implement the `sort` strategy and `conflictPolicy` fields in the initial implementation. That leaves us with `into`, `strategy` and `priority` which should be fairly straight forward.

**Comment by amogh-dambal** [comment]
Hi! Mind if I take this one?

**Comment by Aransh** [comment]
@amogh-dambal not sure who you're asking permission from (I think this is just up for grabs), but in case it's me, go for it! Would be much much appreciated

**Comment by amogh-dambal** [comment]
> not sure who you're asking permission from (I think this is just up for grabs), but in case it's me, go for it!
@Aransh Sweet, thanks! I'll go ahead and assign it to myself. Wanted to check w/ @moolen or another maintainer to make sure this is still a feature open to the community to pull in, but it sounds like it is.

**Comment by amogh-dambal** [comment]
Was able to take a bit of a deeper look at this yesterday. I think I have a pretty good handle on how and where this can be implemented, but I also have a couple of questions about the proposal I wanted to get clarity on. 

### Implementation
To me, the most straightforward implementation of this feature would be to update the [`ExternalSecretRewrite` type](https://github.com/external-secrets/external-secrets/blob/main/apis/externalsecrets/v1beta1/externalsecret_types.go#L282) with a new `ExternalSecretRewriteMerge` that looks something like this:
```
type ExternalSecretRewriteMerge struct {
	Into                  *string `json:"into,omitempty"`
	Strategy          ExternalSecretRewriteMergeStrategy `json:"strategy"`
	ConflictPolicy ExternalSecretRewriteConflictPolicy `json:"conflictPolicy"`
	Sort                 *ExternalSecretRewriteSort `json:"sort,omitempty"`
	Priority            []string `json:"priority,omitempty"`
}
```
then update the [`RewriteMap` function](https://github.com/external-secrets/external-secrets/blob/main/pkg/utils/utils.go#L69) to look something like this:
```
func RewriteMap(operations []esv1beta1.ExternalSecretRewrite, in map[string][]byte) (map[string][]byte, error) {
	out := in
	var err error
	for i, op := range operations {
		if op.Merge != nil {
			out, err = RewriteMerge(*op.Merge, out)
			if err != nil {
				return nil, fmt.Errorf("failed rewriting merge operation[%v]: %w", i, err)
			}
		}
		if op.Regexp != nil {
			out, err = RewriteRegexp(*op.Regexp, out)
			if err != nil {
				return nil, fmt.Errorf("failed rewriting regexp operation[%v]: %w", i, err)
			}
		}
		if op.Transform != nil {
			out, err = RewriteTransform(*op.Transform, out)
			if err != nil {
				return nil, fmt.Errorf("failed rewriting transform operation[%v]: %w", i, err)
			}
		}
	}
	return out, nil
}
```

This would ensure that the merging happens when the `Reconciler` calls `handleFindAllSecrets` [here](https://github.com/external-secrets/external-secrets/blob/main/pkg/controllers/externalsecret/externalsecret_controller_secret.go#L221) and it should be sufficient for the feature as described.

### Questions

- I'm not sure I fully understand the proposed `JSON` merge strategy. With the example above, with secrets that look like:
```
```


cc: @moolen @Skarlso

**Comment by Aransh** [comment]
Hi again @amogh-dambal :)
Appreciate your work on this, I'll try to clarify further.

About the implementation part, should probably leave it to someone more familiar with the code 😅
About the questions:
1. No that's not the proposal, that's the current behavior (or at least closer to it), which doesn't work for my use case. I describe this here - https://github.com/external-secrets/external-secrets/issues/2987#issuecomment-1873511771
The main idea is to provide a directory path in vault, and have external-secrets merge the key/values from all secrets under this directory into a single secret, with multiple keys (only the actual key/values from the secrets, not to place them under their secret name as happens now).
2. With question 1 explained I believe this should clarify this further, let's say I currently have these 2 secrets in vault under /apps/hello-world/credentials:
credentials (provided by devops, generated/managed by terraform)
```
user: test
password: test
```
/apps/hello-world/custom-stuff-controlled-by-developers (manually managed by devs):
```
token: xxx
```
I would like the k8s pod to just set envFrom from a single secret merging all these into one, like:
hello-world-secret:
```
data:
  user: test
  password: test
  token: xxx
```

And now the devs want to overwrite my provided user/password (for some test for example, they want the app to login as a different user), so they update /apps/hello-world/custom-stuff-controlled-by-developers:
```
token: xxx
user: test2
password: test2
```
And now the k8s secret should be:
```
data:
  user: test2
  password: test2
  token: xxx
```

The problem here is how would external-secrets know that the user/password keys on "custom-stuff-controlled-by-developers" takes precedent over those same keys in "credentials"? That's where the "priority" array comes into play.
- If no array provided, by default prioritize/overwrite duplicate keys by secret name (alphabetical)
- If array provided, prioritize keys from the secrets by the order provided (only if these secrets exist)


I hope I was clear enough, if not please feel free to say what isn't clear, and thank you again 🙏🏼

**Comment by sherifabdlnaby** [comment]
Implementing this feature will make it much easier for Platform maintainers to support Apps; I want my ExternalSecret to fetch `<cluster_name>/<application_name>/*` which will combine multiple `<secret-types>` directories into one. I'll be fine with ambiguous merge precedence for keys with duplicate values across the different `<secret-types>` (or just say it'll have alphabetical precedence or whatever)

As a Platform Onwer I'll probably create multiple ExternalSecrets to enforce some type of precedence such as:
-  `<cluster_name>/common/<application_name>/*`
- `<cluster_name>/<region>/<application_name>/*`
Which will have expected precedence (cuz I will mount the K8S secrets in the `envFrom` in the correct order)... and will expect merge between keys of the same prefix ambiguous.

**Comment by sherifabdlnaby** [comment]
Oh... actually I was able to figure this out using [Advanced Templating v2](https://external-secrets.io/main/guides/templating/) feature 🎉 🎉 🎉 🎉 

### Setup
I  created 3 Secrets in Vault:
| Path                                       | Keys and Values                                                          |
|--------------------------------------------|--------------------------------------------------------------------------|
| `/production/service-a/aws`                | `AWS_KEY`: XXXXXXXX <br/> `AWS_secRET`: XXXXXXXX <br/> `AWS_Whtever`: XXXXXXXX |
| `/production/service-a/docusign`           | `DOCU_KEY`: XXXXX <br/> `DOCU_PASS`: 11111                                   |
| `/production/service-a/path-with-conflict` | `AWS_KEY`: value-from-conflict                                             |

---

If you use the follow `ExternalSecret`:
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: service-a
spec:
  secretStoreRef:
    name: vault-backend
    kind: SecretStore
  target:
    name: service-a
    deletionPolicy: Merge
  dataFrom:
    - find:
        path: /production/service-a/
        name:
          regexp: ".*"
```
Notice I used path `/production/service-a/`.

This one outputs the following:
```yaml
_production_service-a__aws: '{"AWS_KEY":"XXXXXXXX","AWS_Whtever":"XXXXXXXX","AWS_secRET":"XXXXXXXX"}'
_production_service-a__docusign: '{"DOCU_KEY":"XXXXX","DOCU_PASS":"11111"}'
_production_service-a__path-with-conflict: '{"AWS_KEY":"value-from-conflict"}'
```

Now let's use the `template` feature to load this output into a Go Template and generate the secret we intended:

```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: service-a
spec:
  secretStoreRef:
    name: vault-backend
    kind: SecretStore
  target:
    name: service-a
    deletionPolicy: Merge
    template:
      engineVersion: v2
      templateFrom:
        - target: Data
          literal: |-
            {{ $data := dict }}
            {{- range $key, $value := . }}
              {{- range $key, $value := (fromJson $value) }}
                {{- $_ := set $data (upper $key) $value -}}
              {{- end -}}
            {{- end -}}
            {{- range $key, $value :=  $data }}
              {{ $key }}: {{ $value }}
            {{- end }}
  dataFrom:
    - find:
        path: /production/service-a/
        name:
          regexp: ".*"
```

This will generate the following Secret:
```yaml
AWS_KEY: value-from-conflict
AWS_SECRET: XXXXXXXX
AWS_WHTEVER: XXXXXXXX
DOCU_KEY: XXXXX
DOCU_PASS: "11111"
```

Notes:
1. I created two paths with same key name inside of them; Because Go Maps are sorted alphabetically when ranging over them you'll consistently have precedence of paths in Alphabetical descending order. So values of keys in `/z` will ALWAYS have precedence over `/a`. In our example `/production/service-a/path-with-conflict/. AWS_KEY` value took precedence over `/production/service-a/aws/  AWS_KEY`.
	2. You can control this behavior by changing the Go Template! You can also fail the entire secret generation if u detect a conflict! 
3. I convert all Keys to Uppercase because I love it this way... you do you... just be mindful that Go Maps and K8S Secrets are case-senstive

**Comment by Aransh** [comment]
@sherifabdlnaby Nice find!
Though I think I'd still rather keep this issue open in hope of an official standard (and simplified) implementation of this.
I'm mostly worried about putting this in our charts and then down the line after other people on our org need to maintain this, as they'll definitely have no clue what's going on here 😂

**Comment by sherifabdlnaby** [comment]
> @sherifabdlnaby Nice find!
> Though I think I'd still rather keep this issue open in hope of an official standard (and simplified) implementation of this.
> I'm mostly worried about putting this in our charts and then down the line after other people on our org need to maintain this, as they'll definitely have no clue what's going on here 😂
> But definitely a solid solution for anyone interested.

Agreed! It's not pretty; it gets even uglier when you template it with Helm! I am also not sure how it'"" handle tricker cases like multiline secrets(will test that today)... 

In my case we maintain a central chart so this logic will only be maintained by a single team and can have extensive documentation... but I would recommend this as something you give to engineers to "copy".

**Comment by riccardomc** [comment]
Hi! Do you mind if I work on this? Looks like there's not much progress and I would like to implement it.

I already have a somewhat working initial implementation in [my fork](https://github.com/external-secrets/external-secrets/compare/main...riccardomc:external-secrets:2987-rewrite-merge)

**Comment by gusfcarvalho** [comment]
@riccardomc Of course!!

Nice to see you back in here 😄😄

**Comment by riccardomc** [comment]
This issue can be closed as it is implemented in #4894 and released in [0.18.1](https://github.com/external-secrets/external-secrets/releases/tag/v0.18.1) now. 🎉

**Comment by Aransh** [comment]
That's amazing! Can't wait to try!

**Comment by fallen-up** [comment]
@riccardomc 
1) are you sure you have created the [correct example](https://github.com/riccardomc/external-secrets/blob/734a46c63b5523201308c1e939d58b40e4aa5bfe/docs/snippets/datafrom-rewrite-merge-empty.yaml)?
```yml
apiVersion: external-secrets.io/v1
kind: ExternalSecret
metadata:
  name: merge-basic-example
spec:
  refreshInterval: 1h
  secretStoreRef:
    name: vault-backend
    kind: SecretStore
  dataFrom:
    - find:
        path: path/to/secrets
        regexp: ".*"
      rewrite:
        - merge: {}
```

```log
Error from server (BadRequest): error when creating "xxx.yaml": ExternalSecret in version "v1" cannot be handled as a ExternalSecret: strict decoding error: unknown field "spec.dataFrom[0].find.regexp"
```
2) How can this manifest be optimized now (when keys in end paths are identical and have different values)?
```yml
spec:
  refreshInterval: 5m
  secretStoreRef:
    kind: ClusterSecretStore
    name: example-namespace
  target:
    name: example-secret
    template:
      engineVersion: v2
      templateFrom:
        - literal: |
            {{- $data := dict -}}
            {{- range $secretPath, $secretValue := . -}}
              {{- $jsonData := fromJson $secretValue -}}
              {{- range $k, $v := $jsonData -}}
                {{- $_ := set $data (printf "%s.%s" $secretPath $k) $v -}}
              {{- end -}}
            {{- end -}}
            {{ toYaml $data | nindent 8 }}
          target: Data
  dataFrom:
    - find:
        path: example-path/path
        name:
          regexp: .*
      rewrite:
        - regexp:
            source: /
            target: .
```

**Comment by gusfcarvalho** [comment]
> 2. How can this manifest be optimized now (when keys in end paths are identical and have different values)?
> 
> spec:
>   refreshInterval: 5m
>   secretStoreRef:
>     kind: ClusterSecretStore
>     name: example-namespace
>   target:
>     name: example-secret
>     template:
>       engineVersion: v2
>       templateFrom:
>         - literal: |
>             {{- $data := dict -}}
>             {{- range $secretPath, $secretValue := . -}}
>               {{- $jsonData := fromJson $secretValue -}}
>               {{- range $k, $v := $jsonData -}}
>                 {{- $_ := set $data (printf "%s.%s" $secretPath $k) $v -}}
>               {{- end -}}
>             {{- end -}}
>             {{ toYaml $data | nindent 8 }}
>           target: Data
>   dataFrom:
>     - find:
>         path: example-path/path
>         name:
>           regexp: .*
>       rewrite:
>         - regexp:
>             source: /
>             target: .

Hi @fallen-up  - regardless if @riccardomc  will reply to you or not, beware of user entitlement. No one involved in this project, maintainers or contributors, are responsible to provide the correct use or help you implement a feature according to your use case. Do what everyone else does - test it in a sandbox environment until you're comfortable enough with your optimization.

User entitlement attitudes have a very low toleration within this current maintaining team.

Also, please move this to a Discussion or a Slack thread. It  is out of scope of this issue how to use the feature that was asked.

**Comment by gusfcarvalho** [comment]
> [@riccardomc](https://github.com/riccardomc)
> 
> 1. are you sure you have created the [correct example](https://github.com/riccardomc/external-secrets/blob/734a46c63b5523201308c1e939d58b40e4aa5bfe/docs/snippets/datafrom-rewrite-merge-empty.yaml)?
> 
> apiVersion: external-secrets.io/v1
> kind: ExternalSecret
> metadata:
>   name: merge-basic-example
> spec:
>   refreshInterval: 1h
>   secretStoreRef:
>     name: vault-backend
>     kind: SecretStore
>   dataFrom:
>     - find:
>         path: path/to/secrets
>         regexp: ".*"
>       rewrite:
>         - merge: {}
> ```
> Error from server (BadRequest): error when creating "xxx.yaml": ExternalSecret in version "v1" cannot be handled as a ExternalSecret: strict decoding error: unknown field "spec.dataFrom[0].find.regexp"

And yes, the correct manifest should be:
```
 apiVersion: external-secrets.io/v1
 kind: ExternalSecret
 metadata:
   name: merge-basic-example
 spec:
   refreshInterval: 1h
   secretStoreRef:
     name: vault-backend
     kind: SecretStore
   dataFrom:
     - find:
         path: path/to/secrets
         name:
           regexp: ".*"
       rewrite:
         - merge: {}
```
Fixed in https://github.com/external-secrets/external-secrets/pull/4960

**Comment by riccardomc** [comment]
Hey @fallen-up, thanks for the feedback and for spotting the wrong example. And thanks @gusfcarvalho for fixing it so fast.

If by "optimization" of your example you mean reducing the lines of code for your `ExternalSecret` then let's take a look together. Given a provider with the following secrets:
```yaml
apiVersion: external-secrets.io/v1
kind: ClusterSecretStore
metadata:
  name: fake
spec:
  provider:
    fake:
      data:
      - key: "example-path/path/database"
        value: '{"username": "username1", "password": "password1"}'
        version: "v1"
      - key: "example-path/path/s3"
        value: '{"accesskey": "accesskey_s3", "secretkey": "secretkey_s3"}'
        version: "v1"
      - key: "example-path/path/minio"
        value: '{"accesskey": "accesskey_minio", "secretkey": "secretkey_minio"}'
```

If you use the following `ExternalSecret`:
```yaml
apiVersion: external-secrets.io/v1
kind: ExternalSecret
metadata:
  name: external-secret-merge
spec:
  secretStoreRef:
    name: fake
    kind: ClusterSecretStore
  dataFrom:
    - find:
        path: example-path/path
        name:
          regexp: ".*"
      rewrite:
        - merge:
            conflictPolicy: Ignore
```

The resulting `Secret` will be:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: external-secret-merge
type: Opaque
data:
  accesskey: YWNjZXNza2V5X3Mz # accesskey_s3
  password: cGFzc3dvcmQx # password1
  secretkey: c2VjcmV0a2V5X3Mz # secretkey_s3
  username: dXNlcm5hbWUx # username1
```

What happens is that the key/values in the `example-path/path/s3` secret end up in the resulting `Secret` data because they are evaluated last due to the alphabetical order. Notice that the `conflictPolicy` _needs_ to be set to `Ignore` for this to work, otherwise conflicts will cause errors.

I guess it _feels_ more readable and more usable by an application than the template version. But it doesn't do exactly the same thing. In fact, the template resulting data would be:
```yaml
data:
  example-path_path_database.password: cGFzc3dvcmQx
  example-path_path_database.username: dXNlcm5hbWUx
  example-path_path_minio.accesskey: YWNjZXNza2V5X21pbmlv
  example-path_path_minio.secretkey: c2VjcmV0a2V5X21pbmlv
  example-path_path_s3.accesskey: YWNjZXNza2V5X3Mz
  example-path_path_s3.secretkey: c2VjcmV0a2V5X3Mz
```

So, ultimately, depends on what you need.
